README
Welcome to OpenBIOS
-------------------

OpenBIOS is a free, portable implementation of IEEE 1275-1994 
(Open Firmware). Find detailed information about OpenBIOS at 
http://www.openbios.org/

What is OpenBIOS?
-----------------

OpenBIOS can replace your system firmware (BIOS) partly or completely. It
can also be used as a bootloader to create an Open Firmware compatible
interface between legacy firmware and an operating system.

This is achieved by a modular concept that consists of a portable Forth
kernel and three interfaces for user interaction, device initialization
and client (operating system) control.

While far not all possible applications of OpenBIOS are implemented yet,
a lot of functionality is already there. OpenBIOS can be used to enhance
LinuxBIOS (http://www.linuxbios.org), or be booted from any multiboot
capable bootloader to bring Open Firmware to your machine. OpenBIOS can
also be used when an operating system is already running. It provides
the needed OpenFirmware functionality to MOL (MacOnLinux) to boot MacOS
9 and X on PPC machines, as well as Linux (all supported platforms)

OpenBIOS build options
---------------------

   config/scripts/switch-arch <platform>  - build for specified platform
   					    Look in config/example for
					    platforms.

   make            - build all configured binaries

   make run        - run unix example.

   
How OpenBIOS works
------------------

 The OpenBIOS forth core is split into a forth kernel written in portable 
 C and a forth dictionary which operated on by the kernel.

 When building the forth core, you get different versions of
 the forth kernel: 

 * a unix executable program

   - to execute a forth dictionary from a file. This can be used for
     easily testing and developing OpenBIOS on a unix host.

   - to create a dictionary file. Such a dictionary file sets up
     all of the forth language. Primitives are indexed to save relocations.

     The default is to create a forth dictionary forth.dict from
     forth/start.fs. This file includes all of the basic forth language
     constructs from forth/bootstrap.fs and starts the interpreter.

     To achieve this, the hosted unix version contains a basic set of
     forth words coded in C that allow creating a full dictionary.

 * a varying number of target specific binaries. On x86 you can start 
   openbios for example from GRUB or LinuxBIOS. They are all based on
   the same forth engine consisting of a dictionary scheduler, primitive 
   words needed to build the forth environment, 2 stacks and a simple 
   set of console functions. These binaries can not be started directly
   in the unix host environment.

Requirements
------------
 * gcc
 * gnu make
 * grub or any other multiboot loader to run the multiboot
   binary "openbios.multiboot" with it's module "openbios-<platform>.dict"
 * xsltproc
 
Building & Usage
----------------

 * make

   this builds "openbios.multiboot", the standalone image and "openbios-unix", 
   the hosted image. Additionally it creates a forth dictionary
   file from forth/start.fs. All generated files are written to 
   the absolute directory held by the variable BUILDDIR, which defaults
   to obj-[platform]. Some compile time parameters can be tweaked in
   include/config.h
   
 * use "openbios-unix" to create a forth dictionary on your own:
   $ obj-x86/openbios-unix -Iforth start.fs
   creates the file forth.dict from forth source forth/start.fs.

 * use "openbios-unix" to run a created dictionary: 
   $ obj-x86/openbios-unix obj-x86/openbios-unix.dict
   This is useful for testing
 
 * booting openbios
   You can boot openbios i.e. in grub. Add the following lines to
   your menu.lst:

    title openbios
      kernel (hd0,2)/boot/openbios.multiboot
      module (hd0,2)/boot/openbios-x86.dict

   Note: change (hd0,2) to the partition you copied the openbios image and
   openbios-x86.dict to.

   To boot OpenBIOS from LinuxBIOS/etherboot, you can either use
   "openbios-plain.elf" or "openbios-builtin.elf":

   - openbios-plain.elf is the pure kernel that loads the dictionary from a 
     hardcoded address in flash memory (0xfffe0000)

   - openbios-builtin.elf also includes the dictionary directly so that it
     can be easily used from etherboot or the LinuxBIOS builtin ELF
     loader without taking care of the dictionary

CREDITS
-------
OpenBIOS was developed by Stefan Reinauer, Samuel Rydh and Patrick Mauritz.
The OpenBIOS IDE driver was written by Jens Axboe.
For license details on this piece of software, see Documentation/COPYING.


If you have patches, questions, comments, feel free to contact the OpenBIOS
mailinglist.

Regards,
     the OpenBIOS team



toke/ChangeLog
*** 0.6.10 (release) - 2005/10/15 ***************************************
 
 * fix hex parsing in strings. Reported by Arti Itra.

*** 0.6.9 (release) - 2005/10/05 ****************************************

 * Bail out with an error if the maximum word count is exceeded.
 * fix bug in endcase that limited fcode output size to 32k. Thanks to
   Prasana Kumar for reporting this.
 * Update FSF address.

*** 0.6.8 (release) - 2005/03/10 ****************************************

 * fcode numbers have been reset to 0x800 when using the 
   fload directive. Fixed.
 * Add makefile to compile as an extra release

******************* 2005/01/02 ******************************************

 * drop endian.h/mconfig.h dependency 

******************* 2004/06/19 ******************************************
 
 * fix off by one error in s"

******************* 2004/06/10 ******************************************
 
 * fix big endian problem in pci header generation.
 * add pci-code-revision (same as pci-revision)

******************* 2004/04/23 ******************************************

 * remove silly warning again
 * other minor changes and bug fixes have been included

******************* 2003/12/13 ******************************************

 * fix emission of device class in pci headers
 * add encode-file keyword
 * add interpose fcode

*** 0.6 (release) - 2003/11/28 ******************************************

 * make lots of functions and variables static
 * update email address
 * handle big output files right
 * clean up pci header code.
 * add pci-header-end, set-rev-level for compatibility 
   to firmworks tokenizer.
 * fix string escape handling

******************* 2003/08/16 ******************************************

 * fit Makefile into openbios build process (allows out of place build)
 * fix address error in emit.c
 
******************* 2002/07/18 ******************************************

 * implement FCODE-VERSION1 and FCODE-VERSION3 (used by SUN's tokenizer)
 * fix uneven embedded hex message for " "(00 FF FF FF FF FF FF 00)"

******************* 2002/07/16 ******************************************

 * implement PCI-REVISION, NOT-LAST-IMAGE
 * implement FCODE-TIME and FCODE-DATE
 * fix header fixup
 * add dependencies to Makefile
 * align to 512byte if we have a PCI header
 * use 128k output buffer per default.
 * rename example suffix to .fs
 * add version1.fs, date.fs, pciexample.fs to examples

******************* 2002/07/15 ******************************************

 * fix CHAR/[CHAR] warnings.
 * remove some debugging from stack.c
 * some more line number fixes.
 * compile with -pedantic on gcc systems
 * fix header generation. 

******************* 2002/07/12 ******************************************

 * move offs16 to scanner.c
 * generate version1 header correctly and switch
   to 8bit offsets when it's detected.
 * let offset16 emit it's fcode# as well
 * implement PCI-HEADER


******************* 2002/07/11 ******************************************

 * add CHAR and [CHAR]
 * change BUFFER to BUFFER:
 * add H.7 renamed fcode words as macros
 * add sample scsi driver package from IEEE 1275-1994
 * new function emit_token supersedes emit_fcode(lookup_token())

*** 0.4 (release) - 2002/05/24 ******************************************

 * remove if..then from abort" (conditionally) 
 * fix linked list juggling in case..endcase constructs

******************* 2002/05/22 ******************************************

 * fix get_word() not to read beyound end of strings
 * more adequate implementation of abort"

******************* 2002/03/27 ******************************************
 
 * add name and endif macro
 * fix typo in finish-device, new-device
 * use new number parsing function to handle dots.
 * do proper handling of hex values embedded in strings.
 * reset fcode word number counter when compiling multiple files.

******************* 2002/03/21 ******************************************

 * some IEEE 1275 compliance changes (forbid definitions in definitions)
 * fix macros
 * fix hex, octal, decimal inside colon definitions

*** 0.2a (bugfix) - 2002/03/21 ******************************************

 * fix off by one bug in string escape codes
 * fix line number calculation
 * do proper debugging output

*** 0.2 (release) - 2002/03/20 ******************************************

 * implemented (most of the) missing control words
 * implemented next-fcode tokenizer directive
 * implemented line numbers for errors and warnings.
 * improved string handling.
 * stop on errors by default.
 * use case insensitive dictionary lookup.
 * lots of debug made compile time conditional

*** 0.1 (release) - 2002/03/04 ******************************************

 * initial release



libc/misc.c
/* 
 *   Creation Date: <2002/10/19 21:05:07 samuel>
 *   Time-stamp: <2002/10/22 22:29:18 samuel>
 *   
 *	<misc.c>
 *	
 *	Miscellaneous
 *   
 *   Copyright (C) 2002, 2003 Samuel Rydh (samuel@ibrium.se)
 *   
 *   This program is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU General Public License
 *   as published by the Free Software Foundation
 *   
 */

#include "openbios/config.h"
#include "libc/string.h"

int errno;

void
qsort( void *base, size_t nmemb, size_t size, int (*compar)(const void *, const void*) )
{
	int worked, i, j;

	/* even more inefficient than the glibc variant :-) */
	do {
		char *p = base;
		worked = 0;
		for( i=0; i<nmemb-1; i++, p+= size ) {
			if( compar( p, p + size ) > 0 ) {
				worked = 1;
				for( j=0; j<size; j++ ) {
					char ch = p[j];
					p[j] = p[j+size];
					p[j+size] = ch;
				}
			}
		}
	} while( worked );
}


long int
strtol( const char *nptr, char **endptr, int base )
{
	int sum, n, sign=1;
	while( isspace(*nptr) )
		nptr++;

	if( *nptr == '-' || *nptr == '+' )
		sign = (*nptr++ == '-') ? -1 : 1;

	if( base == 16 || base == 0) {
		if( !base )
			base = (nptr[0] == '0')? 8 : 10;
		if( nptr[0] == '0' && nptr[1] == 'x' ) {
			nptr += 2;
			base = 16;
		}
	}
	for( sum=0 ;; nptr++ ) {
		char ch = *nptr;
		if( !isalnum(ch) )
			break;
		n = isdigit(ch) ? ch - '0' : toupper(ch) - 'A' + 10;
		if( n >= base || n < 0 )
			break;
		sum *= base;
		sum += n;
	}
	if( endptr )
		*endptr = (char*)nptr;

	return sum * sign;
}
